interface/modules/custom_modules/oe-module-faxsms/src/Controller/EtherFaxActions.php

```diff
@@ -561,45 +561,156 @@ public function formatFax($encodedFax): string
+
+            // Ensure directory exists
+            $dir = dirname($targetPath);
+            if (!is_dir($dir) && !mkdir($dir, 0770, true)) {
+                return json_encode(['success' => false, 'message' => xlt('Failed to create directory')]);
+            }
+
+            // Write atomically
+            $tmp = tempnam($dir, 'fax_');
+            if ($tmp === false) {
+                return json_encode(['success' => false, 'message' => xlt('Failed to create temp file')]);
+            }
+            $bytes = file_put_contents($tmp, $decoded, LOCK_EX);
+            if ($bytes === false) {
+                @unlink($tmp);
+                return json_encode(['success' => false, 'message' => xlt('Failed to write file')]);
+            }
+            // Move to destination
+            if (!@rename($tmp, $targetPath)) {
+                @unlink($tmp);
+                return json_encode(['success' => false, 'message' => xlt('Failed to finalize file')]);
+            }
+            @chmod($targetPath, 0660);
+
+            $response['success'] = true;
+            $response['url'] = $targetPath;
+            return json_encode($response);
+        } elseif ($action === 'setup') {
+            // Setup without content should not create files; just acknowledge if path is allowed
+            $response['success'] = true;
+            $response['url'] = $targetPath;
+            return json_encode($response);
         }
 
-        if (!empty($content) && $action == 'setup') {
-            $decodedContent = base64_decode((string) $content);
-            if (file_put_contents($where, $decodedContent) !== false) {
-                $response['success'] = true;
-                $response['url'] = $where;
+        // Default: unsupported action
+        return json_encode(['success' => false, 'message' => xlt('Unsupported action')]);
+    }
+
+    /**
+     * Normalize a filesystem path without resolving symlinks from user input.
+     */
+    private function normalizePath(string $path): string
+    {
+        // If the path is relative, anchor it under allowed baseDir
+        if (!str_starts_with($path, DIRECTORY_SEPARATOR)) {
+            $path = rtrim((string) $this->baseDir, DIRECTORY_SEPARATOR) . DIRECTORY_SEPARATOR . ltrim($path, DIRECTORY_SEPARATOR);
+        }
+        // Collapse .. and .
+        $parts = [];
+        foreach (explode(DIRECTORY_SEPARATOR, $path) as $segment) {
+            if ($segment === '' || $segment === '.') {
+                continue;
+            }
+            if ($segment === '..') {
+                array_pop($parts);
             } else {
-                $response['message'] = 'Failed to write file';
+                $parts[] = $segment;
             }
-        } elseif ($action == 'setup') {
-            $response['success'] = true;
-            $response['url'] = $where;
         }
+        $prefix = DIRECTORY_SEPARATOR;
+        if (preg_match('/^[A-Za-z]:\\/', $path) === 1) { // Windows drive
```

Comment:
I think that regex is fragile, dealing with backslashes. This is more verbose, but maybe easier to maintain… WDYT?

```suggestion
        $isWindowsDrive = strlen($path) >= 3 && ctype_alpha($path[0]) && $path[1] === ':' && ($path[2] === DIRECTORY_SEPARATOR);
        if ($isWindowsDrive) {
```